module/zfs/dsl_scan.c

```diff
@@ -835,6 +865,13 @@ dsl_scan_done(dsl_scan_t *scn, boolean_t complete, dmu_tx_t *tx)
 		 * Let the async thread assess this and handle the detach.
 		 */
 		spa_async_request(spa, SPA_ASYNC_RESILVER_DONE);
+
+		if (dsl_scan_reset_defered(spa->spa_root_vdev, tx) &&
+		    vdev_resilver_needed(spa->spa_root_vdev, NULL, NULL)) {
+			zfs_dbgmsg("starting defered resilver txg=%llu",
+			    (longlong_t)tx->tx_txg);
+			spa_async_request(spa, SPA_ASYNC_RESILVER);
```

Comment:
We mark these vdevs no longer "resilver deferred" in this txg (in the config), but the resilver is started asynchronously.  So I think we could "lose" the resilver if we crash before it starts.  Is that OK because we'll re-check for resilver when we open the pool again?  Assuming that's the case, let's leave a comment here explaining it.

Maybe another way to make this more clear would be to have dsl_scan_reset_defered() not return anything, and only check vdev_resilver_needed() here.  I think that would provide the same behavior.  But that way it doesn't look like we're depending on resilver_deferred to tell us to do a resilver.  I think that the resilver_deferred flag might be more accurately described as "vdev is not part of current resilver".  Because all it needs to do is tell us to ignore this vdev's DTL for resilver purposes (not issue resilver i/os to it, and not clear the DTL when the resilver completes).  (Incidentally, this is why a "participates in resilver" flag would be more natural, but I understand that we don't want to activate a new on-disk feature during all resilvers.)